FWCore/Framework/interface/InputProcessBlockCacheImpl.h

```diff
@@ -0,0 +1,234 @@
+        functors_ = std::move(functors);
+        if (!tokenInfos_.empty()) {
+          processNames_.resize(sizeof...(CacheTypes));
+        }
+      }
+
+      // These are used to fill the CacheTuples
+      // One CacheFiller for each CacheType
+
+      class CacheTuple {
+      public:
+        std::tuple<std::shared_ptr<CacheTypes>...> cacheTuple_;
+      };
+
+      template <std::size_t I>
+      typename std::enable_if<I == sizeof...(CacheTypes), void>::type fillCache(ProcessBlock const&,
+                                                                                CacheTuple const&,
+                                                                                CacheTuple&) {}
+
+      template <std::size_t I>
+          typename std::enable_if < I<sizeof...(CacheTypes), void>::type fillCache(ProcessBlock const& pb,
+                                                                                   CacheTuple const& previousCacheTuple,
+                                                                                   CacheTuple& cacheTuple) {
+        if (pb.processName() == processNames_[I]) {
+          auto const& previousSharedPtr = std::get<I>(previousCacheTuple.cacheTuple_);
+          std::get<I>(cacheTuple.cacheTuple_) = std::get<I>(functors_).func_(pb, previousSharedPtr);
+        }
+        fillCache<I + 1>(pb, previousCacheTuple, cacheTuple);
+      }
+
+      void accessInputProcessBlock(ProcessBlock const& pb) {
+        if (sizeof...(CacheTypes) > 0 && !processNames_.empty()) {
+          CacheTuple cacheTuple;
+          if (caches_.empty()) {
+            CacheTuple firstCacheTuple;
+            fillCache<0>(pb, firstCacheTuple, cacheTuple);
+          } else {
+            CacheTuple const& previousCacheTuple = caches_.back();
+            fillCache<0>(pb, previousCacheTuple, cacheTuple);
+          }
+          caches_.push_back(std::move(cacheTuple));
+        }
+      }
+
+      void clearCaches() { caches_.clear(); }
+      auto cacheSize() const { return caches_.size(); }
+
+    private:
+      template <std::size_t ICacheType, typename CacheType, typename DataType, typename Func>
+      void registerProcessBlockCacheFiller(EDGetTokenT<DataType> const& token, Func&& func) {
+        static_assert(ICacheType < sizeof...(CacheTypes), "ICacheType out of range");
+        processNames_.resize(sizeof...(CacheTypes));
+        tokenInfos_.resize(sizeof...(CacheTypes));
+        if (!tokenInfos_[ICacheType].token_.isUninitialized()) {
+          throw Exception(errors::LogicError)
+              << "registerProcessBlockCacheFiller should only be called once per cache type";
+        }
+
+        tokenInfos_[ICacheType] = TokenInfo{EDGetToken(token), TypeID(typeid(DataType))};
+
+        std::get<ICacheType>(functors_).func_ =
+            std::function<std::shared_ptr<CacheType>(ProcessBlock const&, std::shared_ptr<CacheType> const&)>(
+                std::forward<Func>(func));
```

Comment:
Is the explicit type on the right hand side really needed, or would the following work?
```suggestion
        std::get<ICacheType>(functors_).func_ = std::forward<Func>(func);
```

Reply:
Done. It compiles and it works. I agree it is easier to read using the implicit type conversion.